### PR TITLE
Encryption rename open-rewrite rule

### PR DESCRIPTION
**Pull request by SamBarker**
### Type of change

- Enhancement / new feature

### Description

Building on #1043 as it move the maven co-ordinates I thought I'd experiment with an [Open Rewrite](https://docs.openrewrite.org/) which would automatically perform the migration. This PR also adds a rewrite rule to update Kroxylicious config files.

### Additional Context
To test the rule I added a dependency on the old name to the sample pom and asked it to update it.

```shell
mvn -U org.openrewrite.maven:rewrite-maven-plugin:run \
  -Drewrite.recipeArtifactCoordinates=io.kroxylicious:kroxylicious-migrations:0.5.0-SNAPSHOT \
  -Drewrite.activeRecipes=io.kroxylicious.rewrite.maven.EncryptionModuleToRecordEncryption \
  -pl :kroxylicious-sample

[INFO] Scanning for projects...
Downloading from central: https://repo.maven.apache.org/maven2/org/openrewrite/maven/rewrite-maven-plugin/maven-metadata.xml
Downloaded from central: https://repo.maven.apache.org/maven2/org/openrewrite/maven/rewrite-maven-plugin/maven-metadata.xml (5.6 kB at 15 kB/s)
[INFO] 
[INFO] ----------------< io.kroxylicious:kroxylicious-sample >-----------------
[INFO] Building Sample filters 0.5.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] >>> rewrite-maven-plugin:5.23.3:run (default-cli) > process-test-classes @ kroxylicious-sample >>>
[INFO] 
[INFO] --- maven-enforcer-plugin:3.4.1:enforce (enforce-plugin-versions) @ kroxylicious-sample ---
[INFO] Rule 0: org.apache.maven.enforcer.rules.RequirePluginVersions passed
[INFO] 
[INFO] --- maven-enforcer-plugin:3.4.1:enforce (enforce-java-versions) @ kroxylicious-sample ---
[INFO] Rule 0: org.apache.maven.enforcer.rules.version.RequireJavaVersion passed
[INFO] 
[INFO] --- maven-enforcer-plugin:3.4.1:enforce (enforce-systemtest-isolation) @ kroxylicious-sample ---
[INFO] Rule 0: org.apache.maven.enforcer.rules.dependency.BannedDependencies passed
[INFO] 
[INFO] --- directory-maven-plugin:1.0:highest-basedir (resolve-rootdir) @ kroxylicious-sample ---
[INFO] Highest basedir set to: /Users/sbarker/development/kroxy/kroxylicious
[INFO] 
[INFO] --- formatter-maven-plugin:2.23.0:format (format-sources) @ kroxylicious-sample ---
[INFO] Processed 6 files in 923ms (Formatted: 0, Skipped: 0, Unchanged: 6, Failed: 0, Readonly: 0)
[INFO] 
[INFO] --- impsort-maven-plugin:1.9.0:sort (sort-imports) @ kroxylicious-sample ---
[INFO] Processed 10 files in 00:00.019 (Already Sorted: 10, Needed Sorting: 0)
[INFO] 
[INFO] --- maven-resources-plugin:3.3.1:resources (default-resources) @ kroxylicious-sample ---
[INFO] Copying 1 resource from src/main/resources to target/classes
[INFO] 
[INFO] --- maven-compiler-plugin:3.12.1:compile (default-compile) @ kroxylicious-sample ---
[INFO] Nothing to compile - all classes are up to date.
[INFO] 
[INFO] --- formatter-maven-plugin:2.23.0:format (format-test-sources) @ kroxylicious-sample ---
[INFO] Processed 4 files in 115ms (Formatted: 0, Skipped: 0, Unchanged: 4, Failed: 0, Readonly: 0)
[INFO] 
[INFO] --- maven-resources-plugin:3.3.1:testResources (default-testResources) @ kroxylicious-sample ---
[INFO] Copying 1 resource from src/test/resources to target/test-classes
[INFO] 
[INFO] --- maven-compiler-plugin:3.12.1:testCompile (default-testCompile) @ kroxylicious-sample ---
[INFO] Nothing to compile - all classes are up to date.
[INFO] 
[INFO] <<< rewrite-maven-plugin:5.23.3:run (default-cli) < process-test-classes @ kroxylicious-sample <<<
[INFO] 
[INFO] 
[INFO] --- rewrite-maven-plugin:5.23.3:run (default-cli) @ kroxylicious-sample ---
[INFO] Using active recipe(s) [io.kroxylicious.rewrite.maven.EncryptionModuleToRecordEncryption]
[INFO] Using active styles(s) []
[INFO] Validating active recipes...
[INFO] Project [Sample filters] Resolving Poms...
[WARNING] Failed to download io.kroxylicious:kroxylicious-encryption:0.5.0-SNAPSHOT. Attempted URIs:
  - file:///Users/sbarker/.m2/repository/
[WARNING] Failed to download io.kroxylicious:kroxylicious-encryption:0.5.0-SNAPSHOT. Attempted URIs:
  - file:///Users/sbarker/.m2/repository/
[WARNING] Failed to download io.kroxylicious:kroxylicious-encryption:0.5.0-SNAPSHOT. Attempted URIs:
  - file:///Users/sbarker/.m2/repository/
[WARNING] Failed to download io.kroxylicious:kroxylicious-encryption:0.5.0-SNAPSHOT. Attempted URIs:
  - file:///Users/sbarker/.m2/repository/
[INFO] Project [Sample filters] Parsing source files
[INFO] Running recipe(s)...
[WARNING] Failed to download io.kroxylicious:kroxylicious-bom:0.5.0-SNAPSHOT. Attempted URIs:
  - file:///Users/sbarker/.m2/repository/
[WARNING] Changes have been made to kroxylicious-sample/pom.xml by:
[WARNING]     io.kroxylicious.rewrite.maven.EncryptionModuleToRecordEncryption
[WARNING]         org.openrewrite.maven.ChangeDependencyGroupIdAndArtifactId: {oldGroupId=io.kroxylicious, oldArtifactId=kroxylicious-encryption, newGroupId=io.kroxylicious, newArtifactId=kroxylicious-record-encryption}
[WARNING] Please review and commit the results.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  17.187 s
[INFO] Finished at: 2024-02-29T15:49:56+13:00
[INFO] ------------------------------------------------------------------------
```

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonarcloud warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

**Comment by k-wall** [comment]
It seems like a good idea, but I'd leave it for another time.

If I were to do one thing to aid transition from 0.4.1 to 0.5.0 I'd alias the factory and keep backward compatibility with the old field names in the config.  However I see the delay that'd cause to 0.5.0 outweigh benefits to a small number of users.

File: kroxylicious-migrations/src/main/resources/META-INF/rewrite/rewrite.yaml [hunk 1]
```diff
@@ -7,10 +7,19 @@
...
       newGroupId: io.kroxylicious
       newArtifactId: kroxylicious-record-encryption
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: io.kroxylicious.rewrite.filterConfig.EnvelopeEncryptionFilterToRecordEncryptionFilter
+displayName: "Replace usages of EnvelopeEncryption Filter in config files with the RecordEncryption Filter"
+recipeList:
+  - org.openrewrite.yaml.ChangeValue:
+      keyPath: $.filters.*.type[?(@ == 'EnvelopeEncryption')]
+      value: RecordEncryption
```
Comment (SamBarker) [inline kroxylicious-migrations/src/main/resources/META-INF/rewrite/rewrite.yaml:24]: In theory this should be done with [ChangePropertyValue](https://docs.openrewrite.org/recipes/yaml/changepropertyvalue) however I can't get that to work and this formulation is equivalent anyway.

File: kroxylicious-migrations/src/main/resources/META-INF/rewrite/rewrite.yaml [hunk 2]
```diff
@@ -7,10 +7,19 @@
...
       oldGroupId: io.kroxylicious
       oldArtifactId: kroxylicious-encryption
       newGroupId: io.kroxylicious
       newArtifactId: kroxylicious-record-encryption
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: io.kroxylicious.rewrite.filterConfig.EnvelopeEncryptionFilterToRecordEncryptionFilter
+displayName: "Replace usages of EnvelopeEncryption Filter in config files with the RecordEncryption Filter"
+recipeList:
+  - org.openrewrite.yaml.ChangeValue:
```
Comment (SamBarker) [inline kroxylicious-migrations/src/main/resources/META-INF/rewrite/rewrite.yaml:22]: I'd like a separate recipe here along the lines of 
```yaml
  - org.openrewrite.yaml.ChangeValue:
      keyPath: $.data."config.yaml".filters.*.type[?(@ == 'EnvelopeEncryption')]
      value: RecordEncryption
```
To catch the values in the kubernetes-examples folder but the inner document is being considered as a single literal (I guess)

**Comment by SamBarker** [comment]
> It seems like a good idea, but I'd leave it for another time.

Why would you leave it? Its working, its already done.

> If I were to do one thing to aid transition from 0.4.1 to 0.5.0 I'd alias the factory and keep backward compatibility with the old field names in the config. However I see the delay that'd cause to 0.5.0 outweigh benefits to a small number of users.

Yes, I was planning something similar along with keeping the old maven coordinates for a release cycle.

File: kroxylicious-migrations/src/main/resources/META-INF/rewrite/rewrite.yaml (hunk 2 above)
Comment (k-wall) [reply kroxylicious-migrations/src/main/resources/META-INF/rewrite/rewrite.yaml:22]: It is not only the name of the factory that has been changed.  We've also made changes to io.kroxylicious.kms.provider.hashicorp.vault.config.Config which has affected the configuration.  We already have notes about this in the CHANGELOG.md.

File: kroxylicious-migrations/src/main/resources/META-INF/rewrite/rewrite.yaml [hunk 3]
```diff
@@ -0,0 +1,25 @@
+#
+# Copyright Kroxylicious Authors.
+#
+# Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+#
+
+---
+type: specs.openrewrite.org/v1beta/recipe
```
Comment (k-wall) [inline kroxylicious-migrations/src/main/resources/META-INF/rewrite/rewrite.yaml:8]: shouldn't the rewrite.yaml be part of the module that they relate to?  In this case, the the kroxylicious-record-encryption

File: kroxylicious-migrations/pom.xml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>io.kroxylicious</groupId>
+        <artifactId>kroxylicious-parent</artifactId>
+        <version>0.5.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>kroxylicious-migrations</artifactId>
```
Comment (k-wall) [inline kroxylicious-migrations/pom.xml:20]: don't the rewrites need to end up in a published artefact?  a filter author should be able to develop a filter from the published maven artefacts.  they don't necessarily need a kroxylicious clone.

File: kubernetes-examples/pom.xml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+	<modelVersion>4.0.0</modelVersion>
+	<parent>
+		<groupId>io.kroxylicious</groupId>
+		<artifactId>kroxylicious-parent</artifactId>
+		<version>0.5.0-SNAPSHOT</version>
+	</parent>
+
+	<artifactId>kubernetes-examples</artifactId>
```
Comment (k-wall) [inline kubernetes-examples/pom.xml:19]: the examples currently rely on the scripts, so turning them into a published artefact without including the scripts doesn't make too much sense.

**Comment by SamBarker** [comment]
So we've released 0.5.0 without this so its moot now. But I think the idea is worth revisiting for other API migrations later.

**Pull request closed without merging**